src/core/ext/transport/chttp2/transport/hpack_encoder.cc

```diff
@@ -431,13 +431,27 @@ void HPackCompressor::Framer::Encode(HttpSchemeMetadata,
 }
 
 void HPackCompressor::Framer::Encode(GrpcTraceBinMetadata, const Slice& slice) {
+  const size_t table_entry_size = GrpcTraceBinMetadata::key().size() +
+                                  slice.size() +
+                                  hpack_constants::kEntryOverhead;
+  if (table_entry_size > HPackEncoderTable::MaxEntrySize()) {
+    EmitLitHdrWithBinaryStringKeyNotIdx(
+        Slice::FromStaticString(GrpcTraceBinMetadata::key()), slice.Ref());
+  }
   EncodeIndexedKeyWithBinaryValue(&compressor_->grpc_trace_bin_index_,
-                                  "grpc-trace-bin", slice.Ref());
+                                  GrpcTraceBinMetadata::key(), slice.Ref());
 }
 
 void HPackCompressor::Framer::Encode(GrpcTagsBinMetadata, const Slice& slice) {
+  const size_t table_entry_size = GrpcTagsBinMetadata::key().size() +
```

Comment:
Let's add a function (in hpack_constants.h maybe, next to EntriesForBytes)... something like SizeForEntry(size_t key_length, size_t value_length)